### PR TITLE
[Operator] support instance_norm

### LINKED ISSUE #260: Code Contribution: 【Lv3】【Operator Development】instance_norm
将以下算子贡献至https://github.com/FlagOpen/FlagGems/tree/master/src/flag_gems/ops 目录下

instance_norm

认领方式：
1. Issue中留言评论：「认领人名称+认领」
2. 在贡献者微信群内发送：认领的Issue号例如：「认领+[https://github.com/FlagOpen/FlagGems/issues/xxx]」

### PR DESCRIPTION
**Pull request by ZQPei**
### PR Category

Operator

### Type of Change

New Feature

### Description

Support instance norm with forward and backward, w/wo running stats

### Issue

- Resolves #260

### Progress

- [ ] Change is properly reviewed (1 reviewer required, 2 recommended).
- [x] Change is responded to an issue.
- [x] Change is fully covered by a UT.

### Performance

Environment:
A100 nvlink 80G
pytorch 1.13
cuda 11.8
image

### CODE REVIEW

File: benchmark/test_norm_perf.py [hunk 1]
```diff
@@ -79,6 +79,36 @@ def test_group_and_layer_norm_benchmark(op_name, torch_op, input_fn):
...
+@pytest.mark.parametrize(
+    "op_name, torch_op, input_fn",
+    [
+        pytest.param(
+            "instance_norm",
+            torch.nn.functional.instance_norm,
+            instancenorm_input_fn,
+            marks=pytest.mark.instance_norm,
+        ),
+    ],
```
Comment (StrongSpoon) [inline benchmark/test_norm_perf.py:97]: benchmark for instance_norm could be fused to function test_group_and_layer_norm_benchmark.

File: tests/test_named_ops.py [hunk 2]
```diff
@@ -29,6 +29,8 @@
     "native_group_norm": ("test_accuracy_groupnorm",),
     "layer_norm": ("test_accuracy_layernorm",),
     "native_layer_norm": ("test_accuracy_layernorm",),
+    "instance_norm": ("test_accuracy_instancenorm",),
+    "native_instance_norm": ("test_accuracy_instancenorm",),
```
Comment (StrongSpoon) [inline tests/test_named_ops.py]: test_named_ops.py is deprecated.

File: tests/test_norm_ops.py [hunk 3]
```diff
@@ -136,6 +136,114 @@ def test_accuracy_layernorm(shape, dtype):
...
+        ref_inp,
+        running_mean=ref_running_mean,
+        running_var=ref_running_var,
+        weight=ref_weight,
+        bias=ref_bias,
+        use_input_stats=use_input_stats,
+        momentum=momentum,
+        eps=eps,
+    )
+    (res_out, res_mean, res_rstd) = flag_gems.instance_norm(
```
Comment (StrongSpoon) [inline tests/test_norm_ops.py]: flag_gems.instance_norm is registered into aten library in by lib.impl, and it doesn't need extra verification. so it's okay to call it using torch interface with flag_gems.use_gems context.

File: src/flag_gems/ops/instancenorm.py [hunk 4]
```diff
@@ -0,0 +1,621 @@
...
+        x,
+        weight=None,
+        bias=None,
+        running_mean=None,
+        running_var=None,
+        use_input_stats=False,
+        momentum=0.1,
+        eps=1e-05,
+        cudnn_enable=False,
+    ):
```
Comment (StrongSpoon) [inline src/flag_gems/ops/instancenorm.py:454]: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L3152 
this is the interface of instance_norm in ATen library. our function should keep the same and returns one output tensor.

File: src/flag_gems/ops/instancenorm.py [hunk 5]
```diff
@@ -0,0 +1,621 @@
...
+            4,
+            5,
+        ], f"x.shape should be [B, C, N] or [B, C, H, W] or [B, C, H, W, L], but got {x.shape}"
+        B, C = x.shape[:2]
+        N = math.prod(x.shape[2:])
+        M = x.numel() // N
+
+        x = x.contiguous()
+        weight = weight.contiguous()
+        bias = bias.contiguous()
```
Comment (StrongSpoon) [inline src/flag_gems/ops/instancenorm.py]: consider the case in which weight or bias is None

Comment (StrongSpoon) [reply src/flag_gems/ops/instancenorm.py]: this is also missing for LayerNorm and groupnorm. we will fix it soon.

File: src/flag_gems/ops/instancenorm.py [hunk 6]
```diff
@@ -0,0 +1,621 @@
...
+        if has_running_stats:
+            assert (
+                N > 1
+            ), f"Expected more than 1 spatial element when training, got input size {x.shape}"
+            assert (
+                running_mean is not None and running_var is not None
+            ), "running_mean and running_var should not both be None"
+            assert (
+                running_mean.shape == running_var.shape and running_mean.shape[0] == C
+            ), f"running_mean and running_var should have shape as {[C,]}"
```
Comment (StrongSpoon) [inline src/flag_gems/ops/instancenorm.py:484]: why only require the 0 dim of running_mean's shape?

File: tests/test_norm_ops.py [hunk 7]
```diff
@@ -136,6 +136,114 @@ def test_accuracy_layernorm(shape, dtype):
...
+    if use_input_stats:
+        ref_mean = torch.mean(ref_inp, dim=list(range(2, inp.ndim)))
+        ref_var = torch.var(ref_inp, dim=list(range(2, inp.ndim)), correction=0)
+        ref_rstd = torch.rsqrt(ref_var + eps)
+    else:
+        ref_mean = ref_running_mean.reshape(-1, C).repeat(B, 1)
+        ref_var = ref_running_var.reshape(-1, C).repeat(B, 1)
+        ref_rstd = torch.rsqrt(ref_var + eps)
+    gems_assert_close(res_mean, ref_mean, res_mean.dtype)
+    gems_assert_close(res_rstd, ref_rstd, res_rstd.dtype)
```
Comment (StrongSpoon) [inline tests/test_norm_ops.py]: not necessary.

File: src/flag_gems/ops/instancenorm.py [hunk 8]
```diff
@@ -0,0 +1,621 @@
...
+    pid = tl.program_id(0)
+    m_mask = pid < M
+    c_offsets = pid % C
+
+    n_offsets = tl.arange(0, TILE_N)
+    mask = n_offsets < N
+
+    x = tl.load(in_ptr + pid * N + n_offsets, mask, other=0.0).to(tl.float32)
+    m = tl.load(running_mean_ptr + c_offsets, mask=m_mask)
+    var = tl.load(running_var_ptr + c_offsets, mask=m_mask)
```
Comment (StrongSpoon) [inline src/flag_gems/ops/instancenorm.py:254]: why not update running_mean and running_var here?

**Review by StrongSpoon — COMMENTED** [review]
thanks for your contribution! please modify the code or respond to the comments.

File: src/flag_gems/ops/instancenorm.py (hunk 8 above)
Comment (ZQPei) [reply src/flag_gems/ops/instancenorm.py:254]: - if `use_input_stats == False`,  `instance_norm_use_running_stats_kernel` uses `running_mean` and `running_var` as mean and var, and there is no need to update `running_mean` as well as `running_var`.
- if `use_input_stats == True`, `update_running_stats_kernel` update `running_mean` and `running_var` at the end of instancenorm

File: src/flag_gems/ops/instancenorm.py (hunk 6 above)
Comment (ZQPei) [reply src/flag_gems/ops/instancenorm.py:484]: because `running_mean's` shape is [C] in instance-norm, same with pytorch

File: src/flag_gems/ops/instancenorm.py (hunk 5 above)
Comment (ZQPei) [reply src/flag_gems/ops/instancenorm.py]: Okay

File: tests/test_norm_ops.py [hunk 9]
```diff
@@ -136,6 +136,115 @@ def test_accuracy_layernorm(shape, dtype):
...
+        ref_inp,
+        running_mean=ref_running_mean,
+        running_var=ref_running_var,
+        weight=ref_weight,
+        bias=ref_bias,
+        use_input_stats=use_input_stats,
+        momentum=momentum,
+        eps=eps,
+    )
+    res_out = flag_gems.instance_norm(
```
Comment (StrongSpoon) [inline tests/test_norm_ops.py]: with flag_gems.use_gems():
    res_out = torch.nn.functional.instance_norm(...)

File: benchmark/test_norm_perf.py [hunk 10]
```diff
@@ -70,9 +88,15 @@ def layernorm_input_fn(shape, dtype, device):
             layernorm_input_fn,
             marks=pytest.mark.layer_norm,
         ),
+        pytest.param(
+            "instance_norm",
+            torch.instance_norm,
```
Comment (StrongSpoon) [inline benchmark/test_norm_perf.py:93]: torch.nn.functional.instance_norm

File: src/flag_gems/ops/instancenorm.py [hunk 11]
```diff
@@ -0,0 +1,664 @@
...
+                        bias,
+                        mean,
+                        rstd,
+                        M,
+                        N,
+                        C,
+                        eps,
+                        HAS_WEIGHT_BIAS=has_weight_bias,
+                    )
+                if has_running_stats and use_input_stats:  # update running stats
```
Comment (StrongSpoon) [inline src/flag_gems/ops/instancenorm.py:552]: redundant condition

File: tests/test_norm_ops.py [hunk 12]
```diff
@@ -136,6 +136,115 @@ def test_accuracy_layernorm(shape, dtype):
...
+        (res_in_grad, res_weight_grad, res_bias_grad) = torch.autograd.grad(
+            res_out, (inp, weight, bias), out_grad
+        )
+    else:
+        (ref_in_grad,) = torch.autograd.grad(ref_out, (ref_inp,), ref_grad)
+        (res_in_grad,) = torch.autograd.grad(res_out, (inp,), out_grad)
+    M = B * C
+    N = inp.numel() // M
+    if use_input_stats:
+        gems_assert_close(res_in_grad, ref_in_grad, dtype, reduce_dim=N)
```
Comment (StrongSpoon) [inline tests/test_norm_ops.py:244]: I thought we should check res_in_grad whether use_input_stats or not.

**Review by StrongSpoon — COMMENTED** [review]
review done.

File: benchmark/test_norm_perf.py (hunk 10 above)
Comment (ZQPei) [reply benchmark/test_norm_perf.py:93]: I found that `torch.nn.functional.instance_norm` uses a different API than `torch.instance_norm`.
This API [here](https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/native_functions.yaml#L3152) is exactly `torch.instance_norm`.

File: src/flag_gems/ops/instancenorm.py (hunk 11 above)
Comment (ZQPei) [reply src/flag_gems/ops/instancenorm.py:552]: Actually No.
`has_running_stats` means `running_mean` and `running_var` is not None.
`use_input_stats` means use `mean` and `rstd` from input, otherwise uses `running_mean` and `running_var` from outside.
Condition when updating running stats, is `has_running_stats` and `use_input_stats` be True simultaneously.

File: tests/test_norm_ops.py (hunk 12 above)
Comment (ZQPei) [reply tests/test_norm_ops.py:244]: Indeed, but I found that if `use_input_stats` is False, the `res_in_grad` is not always correct with pytorch, because the behavior of `use_input_stats` is different at the training and eval stage.
Usually `use_input_stats` being False is at the evals stage.
So maybe there would not be much helpful to check the backward of in_grad?

File: tests/test_norm_ops.py (hunk 9 above)
Comment (ZQPei) [reply tests/test_norm_ops.py]: Sure

**Review by Bowen12992 — COMMENTED** [review]
you can pull the latest code to Pass the CI

File: benchmark/test_norm_perf.py (hunk 10 above)
Comment (StrongSpoon) [reply benchmark/test_norm_perf.py:93]: I thought they are the same.
截屏2024-12-03 10 13 19

File: src/flag_gems/ops/instancenorm.py (hunk 11 above)
Comment (StrongSpoon) [reply src/flag_gems/ops/instancenorm.py:552]: use_input_stats is checked in line 500. so there is no need to check it again.

File: tests/test_norm_ops.py (hunk 12 above)
Comment (StrongSpoon) [reply tests/test_norm_ops.py:244]: when use_input_stats is False, does backward function use running_mean and running_var instead of mean and rstd?

Comment (ZQPei) [reply tests/test_norm_ops.py:244]: > when use_input_stats is False, does backward function use running_mean and running_var instead of mean and rstd?

No, I tried that way, but got mismatch with pytorch

Comment (StrongSpoon) [reply tests/test_norm_ops.py:244]: got it.

**Review by StrongSpoon — APPROVED** [review]
lgtm
